### ISSUE TITLE
Factor out common PostgreSQL code

### ISSUE THREAD

**Issue by axman6** [tathougies/beam#118]
I'm not completely sure if this is possible at this point, but I was wondering how difficult it would be to factor out the common PostgreSQL syntax generation logic from the PostgreSQL-simple specific parts. I'm interested in seeing how difficult making a `hasql`/`postgresql-binary` backend would be, and can't actually tell how closely the syntax generation is tied to the `postgresql-simple` specific code.

If the SQL generation code could be factored out, that would probably make the implementation of a hasql backend much simpler (and any other postgres based library too I'd imagine), since hasql is mostly built around writing your own queries and providing serialisers and deserialisers for the specific types being passed as parameters and read from query results.

**Comment by tathougies** [comment]
This should already be possible. Just implement a new instance of `MonadBeam`, 

```
instance MonadBeam PgCommandSyntax PgBinary Connection PgBinaryM
```

where `PgBinary` is a new type with a `BeamBackend` instance and `PgBinaryM` is a new type defining the monad in which the `MonadBeam` operations are supported (this can probably be a wrapper around `PgM` too, since that's just a free monad). `PgBinary` would have to have associated `FromBackendRow` operations, to do the deserialization from the postgres binary data types.

The only gotcha here is that you'll still be relying on the (text) encoding from postgresql-simple when beam generates the query text. However, I'm planning on making the postgres syntax more generic in this regard. I just started [`beam-mysql`](https://github.com/tathougies/beam-mysql), where I'm exploring some [more efficient ways](https://github.com/tathougies/beam-mysql/blob/master/Database/Beam/MySQL/Syntax.hs#L24) of encoding the beam syntaxes, and I think the approach there would allow you a very straightforward way to customize the value encoding.

Do feel free to take a stab at it. Happy to help out on IRC, and implement whatever core changes may be necessary. Having two backends with the same syntax is something I've tried to design `beam-core` to handle without issue.
